src/utils.rs

```diff
@@ -0,0 +1,228 @@
+        self.into_bigint().into_be_bytes32()
+    }
+}
+
+impl IntoBEBytes32 for Fq {
+    fn into_be_bytes32(self) -> [u8; 32] {
+        self.into_bigint().into_be_bytes32()
+    }
+}
+
+impl IntoBEBytes32 for u64 {
+    fn into_be_bytes32(self) -> [u8; 32] {
+        let be = self.to_be_bytes();
+        let mut arr = [0u8; 32];
+        arr[24..].copy_from_slice(&be);
+        arr
+    }
+}
+
+pub(crate) fn read_u64(data: &[u8]) -> Result<(u64, &[u8]), ()> {
+    let value = u64::from_be_bytes(data[..8].try_into().map_err(|_| ())?);
+    Ok((value, &data[8..]))
+}
+
+pub(crate) fn read_u256(bytes: &[u8]) -> Result<U256, ()> {
+    <&[u8; 32]>::try_from(bytes)
+        .map_err(|_| ())
+        .map(IntoU256::into_u256)
+}
+
+// Parse point in G1.
+pub(crate) fn read_g1<H: CurveHooks>(data: &[u8]) -> Result<(G1<H>, &[u8]), ()> {
+    if data.len() < 64 {
+        return Err(());
+    }
+
+    let x = Fq::from_bigint(read_u256(&data[0..32])?).ok_or(())?;
+    let y = Fq::from_bigint(read_u256(&data[32..64])?).ok_or(())?;
+
+    // If (0, 0) is given, we interpret this as the point at infinity:
+    // https://docs.rs/ark-ec/0.5.0/src/ark_ec/models/short_weierstrass/affine.rs.html#212-218
+    if x == Fq::ZERO && y == Fq::ZERO {
+        return Ok((G1::zero(), &data[64..]));
+    }
+
+    let point = G1::new_unchecked(x, y);
+
+    // Validate point
+    if !point.is_on_curve() {
+        return Err(());
+    }
+    // This cannot happen for G1 with the BN254 curve.
+    // if !point.is_in_correct_subgroup_assuming_on_curve() {
+    //     return Err(());
+    // }
+
+    Ok((point, &data[64..]))
+}
+
+// Parse point in G2.
+pub(crate) fn read_g2<H: CurveHooks>(data: &[u8]) -> Result<G2<H>, ()> {
+    if data.len() != 128 {
+        return Err(());
+    }
+
+    // Read in reverse order (i.e., imaginary part before real part) to match
+    // Solidity's encoding:
+    // https://eips.ethereum.org/EIPS/eip-197#encoding
+    let x_c1 = read_fq_util(&data[0..32]).expect("Parsing the SRS should always succeed!");
+    let x_c0 = read_fq_util(&data[32..64]).expect("Parsing the SRS should always succeed!");
+    let y_c1 = read_fq_util(&data[64..96]).expect("Parsing the SRS should always succeed!");
+    let y_c0 = read_fq_util(&data[96..128]).expect("Parsing the SRS should always succeed!");
+
+    let x = Fq2::new(x_c0, x_c1);
+    let y = Fq2::new(y_c0, y_c1);
+
+    Ok(G2::<H>::new(x, y))
+}
+
+// Utility function for parsing points in G2.
+pub(crate) fn read_fq_util(data: &[u8]) -> Result<Fq, FieldError> {
+    if data.len() != 32 {
+        return Err(FieldError::InvalidSliceLength {
+            expected_length: 32,
+            actual_length: data.len(),
+        });
+    }
+
+    // Convert bytes to limbs manually
+    let mut limbs = [0u64; 4];
+    for (i, chunk) in data.chunks(8).enumerate() {
+        limbs[3 - i] = u64::from_be_bytes(chunk.try_into().unwrap());
+    }
+
+    let bigint = U256::new(limbs);
+
+    Ok(bigint.into_fq())
```

Comment:
Equivalent, more concise implementation

```suggestion
    let mut rchunks_iter = data.rchunks(8);
    let limbs = core::array::from_fn(|_| {
        u64::from_be_bytes(rchunks_iter.next().unwrap().try_into().unwrap())
    });

    Ok(U256::new(limbs).into_fq())
```